### PR TITLE
fix: Escape empty string PathPart

### PR DESCRIPTION
**Pull request by mkmik**
Empty directory names are silently ignored and can lead to very surprising effects
such as directory layouts missing a level. This makes it hard to reason about directory structures.

A sane object store path API should either disallow empty names or deal with them gracefully.

Since we already have to escape file/directory names using the minimum common denominator valid character
set for known cloud providers, it feels quite natural to treat this empty dir/file name problem as encoding problem.

This PR should have enough inline documentation to explain the rationale for the chosen encoding;
if it's not clear, let's improve the inline documentation.

---

Part of the #818 epic; shaving the yak observed when no partition rules are configured and the partition key is an empty string.

### CODE REVIEW

File: object_store/src/path/parts.rs [hunk 1]
```diff
@@ -5,11 +5,17 @@ use super::DELIMITER;
...
+// time.
+const EMPTY: &'static str = "%";
+
 /// The PathPart type exists to validate the directory/file names that form part
 /// of a path.
 ///
-/// A PathPart instance is guaranteed to contain no `/` characters as it can
-/// only be constructed by going through the `try_from` impl.
+/// A PathPart instance is guaranteed to be non-empty and to contain no `/`
+/// characters as it can only be constructed by going through the `from` impl.
```
Comment (mkmik) [inline object_store/src/path/parts.rs:18]: the old comment mentioned `try_from` but that was probably a left-over from a previous impl

**Review by tustvold — APPROVED** [review]
Tentatively approved - I'm personally in the return an error camp, but from an implementation perspective I think this is correct

**Comment by mkmik** [comment]
> Tentatively approved - I'm personally in the return an error camp, but from an implementation perspective I think this is correct

I choose that path initially; but since partition keys can be empty, it just means we'd need to add another layer of encoding when we transform the partition key into a filename, while at same time adding tons of error handling cases that deal with the only possible error of being empty (since all other invalid chars are accepted, including the path separator `/` !!).

This observation made me change my mind and opt for this attempt instead.

**Comment by mkmik** [comment]
> Tentatively approved - I'm personally in the return an error camp, but from an implementation perspective I think this is correct

I choose that path initially; but then I observed two things:

* every character is accepted, even otherwise illegal characters such as the path separator `/`!!
* we use some values such as the partition key as directory names; an empty partition key is valid hence we'd need to add another encoding layer to escape the partition key (and possibly other values); now imagine wrapping this encoding functionality into a helper layer that helps us deal with the object store and voila' you basically have this crate.

**Comment by tustvold** [comment]
Agreed, this is consistent with what is currently implemented. However, intuitively I'd have thought we'd prevent illegal partition parts as opposed to encoding them, so as to allow object storage to be somewhat compatible with other big data tooling, but I don't really have the right context here.

**Review by carols10cents — APPROVED** [review]
This looks good to me from an implementation standpoint; not sure I know enough about every edge case in iox object store logic that might occur though.

**Comment by mkmik** [comment]
> However, intuitively I'd have thought we'd prevent illegal partition parts as opposed to encoding them

my understanding is that the empty string is a perfectly legal partition key

**Review by alamb — APPROVED** [review]
Thanks @mkmik  for fixing the corner case. I also think this is fine, but like @tustvold  would prefer an error if someone tried to create a  `PathPart` from an empty string. My rationale for error is that an empty `PathPart` seems like it would almost always signal a bug / defect elsewhere and I would rather it be flagged / signaled as soon as possible.

**Comment by mkmik** [comment]
I'm in two minds;

on one hand, an API about paths should define what are valid path parts and, like stdlibs of common languages do, disallow the path separator `"/"` and the empty string (well, some forbid empty string, some swallow it silently, but this is besides the point here).

on the other hand we want a helper layer that encodes our domain values safely into object store paths.

Currently we conflate the "path management abstraction" with the "domain value into path codec helper" (as demonstrated by the fact that we currently allow the path separator).

I see two options:

1. embrace this and just escape the empty string as in this PR
2. turn PathPart into a pure objectore path abstraction, that disallows `"/"` and and `""` and only cares of working around cloud vendor specific limitations (e.g. google cloud's `"*"` and non-ascii stuff); and then adding another helper to transcode our domain values into a path, which may (or may not) escape an empty string, depending on the domain of the type to be encoded.

**Comment by alamb** [comment]
> I see two options:

I am happy with either approach and have no strong preference.

**Comment by mkmik** [comment]
ok thanks; let's start with that; things will get clearer as the code calling in object-storage matures

**Pull request merged**